File: 백준/Gold/1005. ACM Craft/ACM Craft.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.PriorityQueue;
import java.util.StringTokenizer;

public class Main {

    public static void main(String[] args) throws Exception {
        BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        int T = Integer.parseInt(bf.readLine());

        for(int t=0; t<T; t++) {
            st = new StringTokenizer(bf.readLine());

            int N = Integer.parseInt(st.nextToken()); // 건물의 개수
            int K = Integer.parseInt(st.nextToken()); // 건설 순서 규칙의 개수

            ArrayList<Integer>[] graph = new ArrayList[N+1];
            for(int i=0; i<=N; i++) {
                graph[i] = new ArrayList<>();
            }

            int[] time = new int[N+1]; // 각 건물 당 건설에 걸리는 시간
            st = new StringTokenizer(bf.readLine());
            for(int i=1; i<=N; i++) {
                time[i] = Integer.parseInt(st.nextToken());
            }

            int[] indegree = new int[N+1];
            for(int i=0; i<K; i++) {
                // 건설 순서 규칙
                st = new StringTokenizer(bf.readLine());
                int u = Integer.parseInt(st.nextToken());
                int v = Integer.parseInt(st.nextToken());

                graph[u].add(v); // v를 건설하기 위해선 u가 건설되어 있어야 함
                indegree[v]++;
            }

            int W = Integer.parseInt(bf.readLine());

            PriorityQueue<int[]> pq = new PriorityQueue<>(
                    (a, b) -> Integer.compare(a[0], b[0])
            ); // [건설 시간, 건물 번호]
            for(int i=1; i<=N; i++) {
                if(indegree[i] == 0) {
                    pq.add(new int[] {time[i], i});
                }
            }

            boolean[] visited = new boolean[N+1];
            int ans = 0;
            while(!pq.isEmpty()) {
                int[] cur = pq.poll();

                if(visited[cur[1]]) {
                    // 이미 지은 건물
                    continue;
                }

                visited[cur[1]] = true;

                if(cur[1] == W){
                    // 지어야 할 건물을 지음
                    ans = cur[0];
                    break;
                }

                for(int next : graph[cur[1]]) {
                    if(!visited[next]) {
                        indegree[next]--;
                        if(indegree[next] == 0) {
                            pq.add(new int[] {cur[0] + time[next], next});
                        }
                    }
                }
            }

            System.out.println(ans);
        }
    }
}

File: 백준/Gold/1005. ACM Craft/ACM Craft.py
import sys
import heapq

input = sys.stdin.readline
T = int(input())

for _ in range(T):
    N, K = map(int, input().split())
    times = [0] + list(map(int, input().split()))

    graph = [[] for i in range(N+1)]
    indegree = [0]*(N + 1)
    for i in range(K):
        a, b = map(int, input().split())
        graph[a].append(b) # b를 짓기 위해선 a를 지어야한다
        indegree[b] += 1

    target = int(input())

    pq = []
    for i in range(1, N+1):
        if indegree[i] == 0:
            heapq.heappush(pq, (times[i], i))
            # 건설에 걸리는 시간으로 최소힙

    cost = 0
    while pq:
        time, cur = heapq.heappop(pq)

        if cur == target:
            cost = time

        for nn in graph[cur]:
            # cur이 가리키는 건물들의 indegree 1감소
            indegree[nn] -= 1

            if indegree[nn] == 0:
                heapq.heappush(pq, (time+times[nn], nn))

    print(cost)